File: src/kotlin/flwr/src/main/java/dev/flower/android/MessageHandler.kt
package dev.flower.android

import flwr.proto.Transport.ClientMessage
import flwr.proto.Transport.ServerMessage
import flwr.proto.Transport.Reason
import flwr.proto.TaskOuterClass.TaskIns
import flwr.proto.TaskOuterClass.TaskRes
import java.lang.IllegalArgumentException


internal fun handle(client: Client, taskIns: TaskIns): Triple<TaskRes, Int, Boolean> {
    val serverMsg = getServerMessageFromTaskIns(taskIns, excludeReconnectIns = false)
        ?: throw NotImplementedError()
    val (clientMsg, sleepDuration, keepGoing) = handleLegacyMessage(client, serverMsg)
    val taskRes = wrapClientMessageInTaskRes(clientMsg)
    return Triple(taskRes, sleepDuration, keepGoing)
}

internal fun handleLegacyMessage(client: Client, serverMsg: ServerMessage): Triple<ClientMessage, Int, Boolean> {
    return when (serverMsg.msgCase.number) {
        1 -> {
            val (disconnectMsg, sleepDuration) = reconnect(serverMsg.reconnectIns)
            Triple(disconnectMsg, sleepDuration, false)
        }
        2 -> {
            Triple(getProperties(client, serverMsg.getPropertiesIns), 0, true)
        }
        3 -> {
            Triple(getParameters(client, serverMsg.getParametersIns), 0, true)
        }
        4 -> {
            Triple(fit(client, serverMsg.fitIns), 0, true)
        }
        5 -> {
            Triple(evaluate(client, serverMsg.evaluateIns), 0, true)
        }
        else -> throw IllegalArgumentException("Unknown Server Message")
    }
}

internal fun reconnect(reconnectMsg: ServerMessage.ReconnectIns): Pair<ClientMessage, Int> {
    var reason = Reason.ACK
    var sleepDuration = 0
    if (reconnectMsg.seconds > 0) {
        reason = Reason.RECONNECT
        sleepDuration = reconnectMsg.seconds.toInt()
    }
    val disconnectRes = ClientMessage.DisconnectRes.newBuilder()
        .setReason(reason)
        .build()
    return Pair(ClientMessage.newBuilder().setDisconnectRes(disconnectRes).build(), sleepDuration)
}

internal fun getProperties(client: Client, getPropertiesMsg: ServerMessage.GetPropertiesIns): ClientMessage {
    val getPropertiesIns = getPropertiesInsFromProto(getPropertiesMsg)
    val getPropertiesRes = client.getProperties(getPropertiesIns)
    return ClientMessage.newBuilder().setGetPropertiesRes(getPropertiesResToProto(getPropertiesRes)).build()
}

internal fun getParameters(client: Client, getParametersMsg: ServerMessage.GetParametersIns): ClientMessage {
    val getParametersIns = getParametersInsFromProto(getParametersMsg)
    val getParametersRes = client.getParameters(getParametersIns)
    return ClientMessage.newBuilder().setGetParametersRes(getParametersResToProto(getParametersRes)).build()
}

internal fun fit(client: Client, fitMsg: ServerMessage.FitIns): ClientMessage {
    val fitIns = fitInsFromProto(fitMsg)
    val fitRes = client.fit(fitIns)
    return ClientMessage.newBuilder().setFitRes(fitResToProto(fitRes)).build()
}

internal fun evaluate(client: Client, evaluateMsg: ServerMessage.EvaluateIns): ClientMessage {
    val evaluateIns = evaluateInsFromProto(evaluateMsg)
    val evaluateRes = client.evaluate(evaluateIns)
    return ClientMessage.newBuilder().setEvaluateRes(evaluateResToProto(evaluateRes)).build()
}

File: src/rust/flwr/src/message_handler.rs
use crate::client;
use crate::flwr_proto as proto;
use crate::serde;
use crate::task_handler;

fn reconnect(reconnect_msg: proto::server_message::ReconnectIns) -> (proto::ClientMessage, i64) {
    let mut reason = proto::Reason::Ack;
    let mut sleep_duration = 0;
    if reconnect_msg.seconds != 0 {
        reason = proto::Reason::Reconnect;
        sleep_duration = reconnect_msg.seconds;
    }

    let disconnect_res = proto::client_message::DisconnectRes {
        reason: reason.into(),
    };

    return (
        proto::ClientMessage {
            msg: Some(proto::client_message::Msg::DisconnectRes(disconnect_res)),
        },
        sleep_duration,
    );
}

fn get_properties(
    client: &dyn client::Client,
    get_properties_msg: proto::server_message::GetPropertiesIns,
) -> proto::ClientMessage {
    let get_properties_res = serde::get_properties_res_to_proto(
        client.get_properties(serde::get_properties_ins_from_proto(get_properties_msg)),
    );

    proto::ClientMessage {
        msg: Some(proto::client_message::Msg::GetPropertiesRes(
            get_properties_res,
        )),
    }
}

fn get_parameters(
    client: &dyn client::Client,
    // get_parameters_msg: proto::server_message::GetParametersIns,
) -> proto::ClientMessage {
    let res = serde::parameter_res_to_proto(client.get_parameters());
    proto::ClientMessage {
        msg: Some(proto::client_message::Msg::GetParametersRes(res)),
    }
}

fn fit(
    client: &dyn client::Client,
    fit_msg: proto::server_message::FitIns,
) -> proto::ClientMessage {
    let res = serde::fit_res_to_proto(client.fit(serde::fit_ins_from_proto(fit_msg)));
    proto::ClientMessage {
        msg: Some(proto::client_message::Msg::FitRes(res)),
    }
}

fn evaluate(
    client: &dyn client::Client,
    evaluate_msg: proto::server_message::EvaluateIns,
) -> proto::ClientMessage {
    let res =
        serde::evaluate_res_to_proto(client.evaluate(serde::evaluate_ins_from_proto(evaluate_msg)));
    proto::ClientMessage {
        msg: Some(proto::client_message::Msg::EvaluateRes(res)),
    }
}

fn handle_legacy_message(
    client: &dyn client::Client,
    server_msg: proto::ServerMessage,
) -> Result<(proto::ClientMessage, i64, bool), &str> {
    match server_msg.msg {
        Some(proto::server_message::Msg::ReconnectIns(reconnect_ins)) => {
            let rec = reconnect(reconnect_ins);
            Ok((rec.0, rec.1, false))
        }
        Some(proto::server_message::Msg::GetParametersIns(_)) => {
            Ok((get_parameters(client), 0, true))
        }
        Some(proto::server_message::Msg::FitIns(fit_ins)) => Ok((fit(client, fit_ins), 0, true)),
        Some(proto::server_message::Msg::EvaluateIns(evaluate_ins)) => {
            Ok((evaluate(client, evaluate_ins), 0, true))
        }
        _ => Err("Unknown server message"),
    }
}

pub fn handle(
    client: &dyn client::Client,
    task_ins: proto::TaskIns,
) -> Result<(proto::TaskRes, i64, bool), &str> {
    let server_msg = task_handler::get_server_message_from_task_ins(&task_ins, false);
    if server_msg.is_none() {
        return Err("Not implemented");
    }
    let (client_msg, sleep_duration, keep_going) =
        handle_legacy_message(client, server_msg.unwrap())?;
    let task_res = task_handler::wrap_client_message_in_task_res(client_msg);
    Ok((task_res, sleep_duration, keep_going))
}
